Includes type forwarding for System.Text.Unicode.Utf8 to the Microsoft.Bcl.Memory library

AlexRadch opened the pull request:
Close #52947

This commit introduces the `System.Text.Unicode.Utf8` type to the `Microsoft.Bcl.Memory` library.

It includes type forwarding for `Utf8` in `Microsoft.Bcl.Memory.Forwards.cs`, updates the documentation in `PACKAGE.md` to include `Utf8` functionality, and adds corresponding test cases in `Microsoft.Bcl.Memory.Tests.csproj`.

The documentation now emphasizes `Utf8` alongside `Index`, `Range`, and `Base64Url`, highlighting its role in converting data between UTF-8 and UTF-16 encodings.

AlexRadch committed:
Add support for System.Text.Unicode.Utf8

This commit introduces the `System.Text.Unicode.Utf8` type to the `Microsoft.Bcl.Memory` library.

It includes type forwarding for `Utf8` in `Microsoft.Bcl.Memory.Forwards.cs`, updates the documentation in `PACKAGE.md` to include `Utf8` functionality, and adds corresponding test cases in `Microsoft.Bcl.Memory.Tests.csproj`.

 The documentation now emphasizes `Utf8` alongside `Index`, `Range`, and `Base64Url`, highlighting its role in converting data between UTF-8 and UTF-16 encodings.

stephentoub commented:
Thank you, but this is not the intent of that work item, which is to actually ship those Utf8 APIs downlevel.

AlexRadch committed:
Enhance UTF-8 and Unicode support in BCL Memory

Updated `PackageDescription` to include "Utf8" support.

Added new `ItemGroup` for conditional compilation of UTF-8
and Unicode handling files for non-net8.0 frameworks.

Modified visibility and implementations in `Ascii.Utility.Helpers.cs`,
`Utf8.cs`, and `Utf8Utility` based on `MICROSOFT_BCL_MEMORY` define.

AlexRadch commented:
> Thank you, but this is not the intent of that work item, which is to actually ship those Utf8 APIs downlevel, e.g. in addition to type forwarding, it also needs an implementation for netstandard.



> Thank you, but this is not the intent of that work item, which is to actually ship those Utf8 APIs downlevel, e.g. in addition to type forwarding, it also needs an implementation for netstandard.

I added Utf8 APIs to Microsoft.Bcl.Memory library. I haven't added tests yet. I'll add them too.

AlexRadch committed:
Enhance Unicode handling in tests and project structure

Updated `Microsoft.Bcl.Memory.Tests.csproj` to include `UnicodeUtility.cs` and removed .NET 8.0 targeting condition.

Modified `Utf8Tests.cs` by adjusting using directives and enhancing the `DecodeHex` method with conditional compilation for .NET 5.0+.

AlexRadch committed:
Add compilation constant for Microsoft BCL Memory

Added a new property `<DefineConstants>$(DefineConstants);MICROSOFT_BCL_MEMORY</DefineConstants>` to the project file to define a new compilation constant for the project.

akoeplinger commented on src/libraries/Microsoft.Bcl.Memory/src/Microsoft.Bcl.Memory.csproj:
```diff
@@ -27,6 +27,18 @@
     <Compile Include="$(CoreLibSharedDir)System\Buffers\Text\Base64Url\Base64UrlValidator.cs" Link="System\Buffers\Text\Base64Url\Base64UrlValidator.cs" />
   </ItemGroup>
 
+  <ItemGroup Condition="!$([MSBuild]::IsTargetFrameworkCompatible('$(TargetFramework)', 'net8.0'))">
```
nit: condition on netframework/netstandard is more readable:
```suggestion
  <ItemGroup Condition="'$(TargetFrameworkIdentifier)' == '.NETFramework' or '$(TargetFrameworkIdentifier)' == '.NETStandard'">
```

akoeplinger replied:
It's just my personal preference, your condition works too but is a bit redundant since it will always be false for net8, net9, net10 since this package only targets net8+ (NetMinimum == net8.0 currently). This is different from Base64Url which we do want to include in the net8 build.

> I don't understand what needs to be fixed.

You can accept the suggestion I posted by clicking on "Commit suggestion", up to you.

teo-tsirpanis replied:
```suggestion
  <ItemGroup Condition="'$(TargetFrameworkIdentifier)' != '.NETCoreApp'">
```

I'd suggest this, if we want to condition on `TargetFrameworkIdentifier`.

AlexRadch replied:
> It's just my personal preference, your condition works too but is a bit redundant since it will always be false for net8, net9, net10 since this package only targets net8+ (NetMinimum == net8.0 currently). This is different from Base64Url which we do want to include in the net8 build.

What if in the future the package is compiled for an older version of .NET, for example for .NET 5+? Then my condition will not need to be changed! It will continue to work correctly, so I would leave it redundant.

jkotas replied:
> What if in the future the package is compiled for an older version of .NET, for example for .NET 5+? 

We are not optimizing for this situation that is very unlikely to happen. We are actively deleting code for no longer supported versions from the repo, https://github.com/dotnet/runtime/pull/109176 is a recent example.

akoeplinger commented on src/libraries/Microsoft.Bcl.Memory/src/PACKAGE.md:
```diff
@@ -11,6 +12,7 @@ This library is not necessary nor recommended when targeting versions of .NET th
 
 * Enables the use of `Index` and `Range` types on older .NET platforms.
 * Provides `Base64Url` encoding, decoding, and validation for URL-safe data processing on older .NET platforms.
+* Provides `Utf8` encoding, decoding, and validation for chunked data between UTF-8 and UTF-16 on older .NET platforms.
```
I'd remove this here since the minimum supported .NET version by the package has built-in Utf8, we only provide it for netframework/netstandard

AlexRadch replied:
I made `Utf8` description similar to  `Base64Url`  description. I don't understand what needs to be fixed.

akoeplinger replied:
`Utf8` is already in net8.0 (minimum .NET platform this package targets) so we only have the type forwarders to the built-in version. This is different from Base64Url which doesn't exist on net8.0 and this package provides it there.

AlexRadch replied:
I changed the package description, as you recommended.

AlexRadch commented:
> Thank you, but this is not the intent of that work item, which is to actually ship those Utf8 APIs downlevel, e.g. in addition to type forwarding, it also needs an implementation for netstandard.

I've also added tests for older .NET platforms.

AlexRadch commented on src/libraries/System.Runtime/tests/System.Runtime.Tests/System/Text/Unicode/Utf8Tests.cs:
```diff
@@ -64,13 +63,45 @@ public static byte[] DecodeHex(ReadOnlySpan<char> inputHex)
         // Don't delete this implementation, as we use it as a reference to make sure the framework's
         // transcoding logic is correct.
-        public static byte[] ToUtf8(Rune rune)
+#if !MICROSOFT_BCL_MEMORY
+        public
+#else
+        private
+#endif
+            static byte[] ToUtf8(Rune rune)
```
In tests for older versions of .Net where there is no Rune class yet, I create it as an internal class, and the compiler complains that the public API cannot reference internal classes.

AlexRadch commented on src/libraries/Microsoft.Bcl.Memory/src/Microsoft.Bcl.Memory.Forwards.cs:
```diff
@@ -3,6 +3,9 @@
 
 [assembly: System.Runtime.CompilerServices.TypeForwardedTo(typeof(System.Index))]
 [assembly: System.Runtime.CompilerServices.TypeForwardedTo(typeof(System.Range))]
+#if NET8_0_OR_GREATER
```
I believe that `#if NET8_0_OR_GREATER` has a clearer indication and does not depend on external conditions that may change.

AlexRadch committed:
Apply suggestions from code review

Co-authored-by: Theodore Tsirpanis <teo@tsirpanis.gr>

jkotas commented on src/libraries/System.Private.CoreLib/src/System/Text/Ascii.Utility.Helpers.cs:
```diff
@@ -60,22 +70,55 @@ internal static uint CountNumberOfLeadingAsciiBytesFromUInt32WithSomeNonAsciiDat
                 // expensive. Instead we'll just change how we perform the shifts.
 
                 // Read first byte
+#if !MICROSOFT_BCL_MEMORY
                 value = BitOperations.RotateLeft(value, 1);
+#else
+                value = (value << 1) | (value >> (32 - 1));
+#endif
```
+1. We prefer polyfills over ifdefs in situations like this.

AlexRadch committed:
Update Microsoft.Bcl.Memory for framework compatibility

- Added polyfill for System.Numerics.BitOperations for .NET Standard 2.0.

AlexRadch commented on src/libraries/System.Private.CoreLib/src/System/Text/Unicode/Utf8.cs:
```diff
@@ -9,7 +9,10 @@
 namespace System.Text.Unicode
 {
-#if SYSTEM_PRIVATE_CORELIB
+#if SYSTEM_PRIVATE_CORELIB || MICROSOFT_BCL_MEMORY
```
I'm afraid to make such changes and let someone else make them as a separate PR.

stephentoub replied:
You can just delete the ifdefs and make this public. I don't know why the ifdefs were here in the first place.

AlexRadch committed:
Space

AlexRadch committed:
Remove MICROSOFT_BCL_MEMORY from project constants

Removed the `DefineConstants` property from the project file, which included the constant `MICROSOFT_BCL_MEMORY`. This change may impact conditional compilation within the project.

AlexRadch committed:
Refactor Utf8Tests for .NET compatibility improvements

teo-tsirpanis commented:
(removing https://github.com/dotnet/runtime/labels/NO-MERGE as the reason it was added has been addressed)

teo-tsirpanis commented:
@dotnet/area-system-memory could you review this? Thanks.

jkotas commented on src/libraries/System.Private.CoreLib/src/System/Text/Unicode/Utf8.cs:
```diff
@@ -815,5 +824,192 @@ private bool Fail()
+        /// the data could not be successfully decoded. This pattern provides convenient automatic U+FFFD substitution of
+        /// invalid sequences while iterating through the loop.
+        /// </remarks>
+        private static OperationStatus DecodeFromUtf8(ReadOnlySpan<byte> source, out uint result, out int bytesConsumed)
```
Can the project include Rune.cs instead of including a copy of the implementation here?

teo-tsirpanis replied:
Done in https://github.com/teo-tsirpanis/dotnet-runtime/commit/79ee05d67f31c0dd7b218c2cf6b2c864b96ce8db in order to unblock work that depends on this, after waiting for one week. Anyone with permissions can push my commit on the PR's branch.

akoeplinger replied:
pushed

teo-tsirpanis replied:
Compile errors fixed in https://github.com/dotnet/runtime/commit/bf6f989f91e88ea75a07e4870faf98a8fb711b88.

akoeplinger replied:
pushed

teo-tsirpanis committed:
Use `Rune.DecodeFromUtf8` on all frameworks.

For downlevel frameworks we add `Rune.cs` to `Microsoft.Bcl.Memory`.

(cherry picked from commit 79ee05d67f31c0dd7b218c2cf6b2c864b96ce8db)

teo-tsirpanis committed:
Fix compile errors.

(cherry picked from commit bf6f989f91e88ea75a07e4870faf98a8fb711b88)

teo-tsirpanis commented:
CI failures are unrelated.

jkotas commented:
I do not have any additional comments. @dotnet/area-system-memory should review and signoff.

teo-tsirpanis commented:
@dotnet/area-system-memory friendly reminder.

stephentoub commented on src/libraries/Microsoft.Bcl.Memory/src/Microsoft.Bcl.Memory.csproj:
```diff
@@ -1,11 +1,11 @@
     <DefineConstants>$(DefineConstants);MICROSOFT_BCL_MEMORY</DefineConstants>
     <IsPackable>true</IsPackable>
-    <PackageDescription>Provides Base64Url, Index and Range types support for .NET Framework and .NET Standard.</PackageDescription>
+    <PackageDescription>Provides Base64Url, Utf8, Index and Range types support for .NET Framework and .NET Standard.</PackageDescription>
```
```suggestion
    <PackageDescription>Provides Base64Url, Utf8, Index, and Range types support for .NET Framework and .NET Standard.</PackageDescription>
```

stephentoub commented on src/libraries/System.Private.CoreLib/src/System/Text/Rune.cs:
```diff
@@ -781,6 +781,7 @@ public int EncodeToUtf8(Span<byte> destination)
 
         public override int GetHashCode() => Value;
 
+#if !MICROSOFT_BCL_MEMORY
```
Why are these `#if !MICROSOFT_BCL_MEMORY` rather than `#if NET` or `#if SYSTEM_PRIVATE_CORELIB`? I'd prefer it if the corelib files didn't need to be aware of Microsoft.Bcl.Memory by name.

stephentoub commented on src/libraries/System.Private.CoreLib/src/System/Text/Rune.cs:
```diff
@@ -1490,7 +1496,11 @@ int IComparable.CompareTo(object? obj)
                 return this.CompareTo(other);
             }
 
+#if SYSTEM_PRIVATE_CORLIB
```
Why is this corelib-specific? The Microsoft.Bcl.Memory package can't include the same resource?

teo-tsirpanis replied:
`Microsoft.Bcl.Memory` does not include any resources at present. And the non-corelib path is highly unlikely to be actually hit.

stephentoub commented on src/libraries/System.Runtime/tests/System.Runtime.Tests/System/Text/Unicode/Utf8Tests.cs:
```diff
@@ -64,13 +63,45 @@ public static byte[] DecodeHex(ReadOnlySpan<char> inputHex)
+            }
+            return result;
+
+            static int FromHex(char c)
+            {
+                if (c >= '0' && c <= '9')
+                {
+                    return c - '0';
+                }
+                else if (c >= 'a' && c <= 'f')
+                {
+                    return c - 'a' + 10;
+                }
+                else
+                {
+                    return c - 'A' + 10;
+                }
+            }
```
Nit:
```suggestion
            static int FromHex(char c) =>
                c >= '0' && c <= '9' ? c - '0' :
                c >= 'a' && c <= 'f' ? c - 'a' + 10 :
                c - 'A' + 10;
```

stephentoub commented on src/libraries/System.Runtime/tests/System.Runtime.Tests/System/Text/Unicode/Utf8Tests.cs:
```diff
@@ -64,13 +63,45 @@ public static byte[] DecodeHex(ReadOnlySpan<char> inputHex)
-        public static byte[] ToUtf8(Rune rune)
+#if NET
+        public
+#else
```
What requires this to be public for NET?

teo-tsirpanis committed:
Address PR feedback.

(cherry picked from commit 445a2325c937eea57f4acea30ed63fae5a2dce13)

teo-tsirpanis commented:
PR feedback addressed in https://github.com/teo-tsirpanis/dotnet-runtime/commit/445a2325c937eea57f4acea30ed63fae5a2dce13.

teo-tsirpanis commented on src/libraries/System.Private.CoreLib/src/System/Text/Unicode/Utf8.cs:
```diff
@@ -9,15 +9,10 @@
-#if SYSTEM_PRIVATE_CORELIB || MICROSOFT_BCL_MEMORY
     /// <summary>
-    /// Provides methods for transcoding between UTF-8 and UTF-16.
+    /// Provides static methods that convert chunked data between UTF-8 and UTF-16 encodings.
```
Documentation copied from Learn.

am11 commented on src/libraries/Microsoft.Bcl.Memory/src/Polyfills/System.Numerics.BitOperations.netstandard20.cs:
```diff
@@ -0,0 +1,75 @@
+        // C# no-alloc optimization that directly wraps the data section of the dll (similar to string constants)
+        // https://github.com/dotnet/roslyn/pull/24621
+
+        private static ReadOnlySpan<byte> TrailingZeroCountDeBruijn => // 32
```
nit: this can be a local in TrailingZeroCount (single use). Slightly better on [netfx as well](https://sharplab.io/#v2:EYLgHgbALANALiAhgZwLYB8ACAmAjAWAChMAGAAk1wDoAlAVwDs4BLVAUyoEkm2AnAewAOAZT4A3ZgGM2yANxEimAMwVsZAMJEA3kTJ6yg3szGI4bCrghkabRABMA8gwA2AT2GDEDADzBXZgD4yABVeRGZnZgYAcwAtPn51fkY4ABE2ACFeOmYAKwYyAF4ggHoSsiVsXX0AbWr9fRISGDISXBbsAA4WkmwOgE4W3FgybBGSJRh6horm0b7RudwAVg7VslwAdh7x7umGpXbR7Y3J0bO8IcGNiB7t/f1sW42F3G7W59wjknXcOZJ+tMALryQjTZRkfjAXJsSRwMgAWQAFABKaY6QgzfSYTaItiofi8VwIxC8ZAAC0QzioAHE2HAbAAzPhsBjSJGhcKRGLxARJFLpLI5fIo0ENAC+RElYJlEJwZFSZAxDQhUJhcMRqPRDz0NnsTjcHi8vn8bCCnIiUTiCX5TEF2TyBUKZDKFSqmKxZDqHs9jX+Ry6PQW2Guww640mOpmSjm2GDS3W2F+JxIuymPt9FQDJ1wFwuR1woeeJHuGd9TyGr3eJE+31+/0BZf0IIUTb0OLxBKJJLJlOpdIZbGZvFZ7It3OtfOSdsyDpFYv00vFQA=).

teo-tsirpanis replied:
I can do it on a subsequent PR.

teo-tsirpanis commented:
Can we merge this?

stephentoub committed:
Merge branch 'main' into BclMemoryUtf8

stephentoub commented:
/ba-g failures are known
